var/spack/repos/builtin/packages/acts/package.py

```diff
@@ -64,13 +65,13 @@ class ActsCore(CMakePackage):
     version('0.08.1', commit='289bdcc320f0b3ff1d792e29e462ec2d3ea15df6')
     version('0.08.0', commit='99eedb38f305e3a1cd99d9b4473241b7cd641fa9')
 
-    # Variants that affect the core ACTS library
+    # Variants that affect the core Acts library
     variant('benchmarks', default=False, description='Build the performance benchmarks')
     variant('examples', default=False, description='Build the examples')
-    variant('tests', default=False, description='Build the unit tests')
     variant('integration_tests', default=False, description='Build the integration tests')
+    variant('unit_tests', default=False, description='Build the unit tests')
```

Comment:
I would actually remove this variant. We don't want `acts+unit_tests` and `acts~unit_tests` to be separate Spack installations, since presumably the same software gets installed. Instead, you can check to see whether or not unit tests are requested via `self.run_tests`. If True, add the flag you need, if not disable it.

Reply:
How would you handle the neighboring `integration_tests` variant though?

- It's not good from an ergonomics point of view if the mechanism for enabling unit tests and integration tests in the package is completely different.
    * Since there is only one `self.run_tests` property, I guess one would need to effectively remove the option of choosing whether to build unit tests or integration tests separately to handle this. On its own, this is pretty reasonable, but...
- ...the `run_tests` class member name suggests that the intent is to not only build tests, but also run them. And there you end up with a problem on your hand:
    * Unit tests can reasonably be run in `Debug` builds, while integration tests cannot (they take several hours without compiler optimizations).
    * You really want to run some tests in `Debug` builds in order to catch firing `assert()`s and get better debugging capabilities to investigate a failing test.
    * It's not good to silently disable some tests depending on the cmake build type, you really want that to be an explicit user choice.

Reply:
I assume the `unit_tests` variant doesn't change what gets installed to the installation prefix, but `integration_tests` does? The dependence on the Debug build type does complicate things. I'll let you make the final decision since you're the maintainer of this package, I'm just here to provide comments based on common packaging pitfalls 😄

Reply:
The integration tests do not get installed to the installation prefix either, from that point of view we're good. Again, if this were only about having a unified variant to build all tests, without running them, I would probably be fine with it.

The main long-term risk that I could think of, with that strategy, is that in the future Acts could want to add integration tests which depend on plugins (and thus on extra dependencies), something which unit tests will never do by design (unit tests are internal to a plugin, you only build the unit tests of the plugins that are enabled).

But that's not a concern today, and either "we'll fix the package when/if it happens" or "you may need to build plugins in order to build tests with Spack" could probably be considered valid ways to handle this risk.

---

Beyond Acts, though, I think in general it would be good for Spack to introduce a more fine-grained way to handle test runners than an on/off switch.

It's not uncommon for software projects to have both "basic" tests which run very quickly and with no extra requirements, and "thorough" tests which take a lot of time to run and can require a special system environment (GPU applications and data acquisition / industrial automation systems come to mind).

Maybe a numerical "test rigor level" (higher is more thorough) would be enough to model this at the Spack level, maybe more is needed, I haven't studied the problem enough to tell...

---

For the time being, I think it's best to leave the `acts` variants as is. I'm pretty confident that users will only enable the package's tests when evaluating a release and turn them off for production, so I don't think it's a **big** problem that packages with test variants enabled don't unify with those that don't have the variants enabled.